<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.31">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Analysis of Canine TCR data from Zuleger et al 2024</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-e1a5c8363afafaef2c763b6775fbf3ca.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-d21829db2b5042f4ae05c2fb504ec7fd.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/quarto-contrib/fontawesome6-6.7.2/all.min.css" rel="stylesheet">
<link href="site_libs/quarto-contrib/fontawesome6-6.7.2/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "?",
    "H"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">SatTCR</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./usage.html"> 
<span class="menu-text">Getting started</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-examples" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Examples</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-examples">    
        <li>
    <a class="dropdown-item" href="./canine_tcr.html">
 <span class="dropdown-text">Canine TCR</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-pre-compiled-reports" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Pre-compiled reports</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-pre-compiled-reports">    
        <li>
    <a class="dropdown-item" href="./cases/canine_tcr/index.html">
 <span class="dropdown-text">Canine TCR</span></a>
  </li>  
    </ul>
  </li>
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://github.com/Ong-Research/SatTCR"> 
<span class="menu-text"><i class="fa-brands fa-github fa-2x" aria-label="github"></i></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-full page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#additional-files-to-download" id="toc-additional-files-to-download" class="nav-link active" data-scroll-target="#additional-files-to-download">Additional files to download</a></li>
  <li><a href="#making-the-samples.csv-file" id="toc-making-the-samples.csv-file" class="nav-link" data-scroll-target="#making-the-samples.csv-file">Making the <code>samples.csv</code> file</a></li>
  <li><a href="#running-the-pipeline" id="toc-running-the-pipeline" class="nav-link" data-scroll-target="#running-the-pipeline">Running the pipeline</a>
  <ul class="collapse">
  <li><a href="#quality-control" id="toc-quality-control" class="nav-link" data-scroll-target="#quality-control">Quality control</a></li>
  <li><a href="#trim-sequences" id="toc-trim-sequences" class="nav-link" data-scroll-target="#trim-sequences">Trim sequences</a></li>
  <li><a href="#clonotype-assembly" id="toc-clonotype-assembly" class="nav-link" data-scroll-target="#clonotype-assembly">Clonotype assembly</a></li>
  <li><a href="#sampling-of-the-sequencing-files-and-saturation-analysis" id="toc-sampling-of-the-sequencing-files-and-saturation-analysis" class="nav-link" data-scroll-target="#sampling-of-the-sequencing-files-and-saturation-analysis">Sampling of the sequencing files and saturation analysis</a></li>
  <li><a href="#building-the-report" id="toc-building-the-report" class="nav-link" data-scroll-target="#building-the-report">Building the report</a></li>
  </ul></li>
  <li><a href="#complete-configuration-file" id="toc-complete-configuration-file" class="nav-link" data-scroll-target="#complete-configuration-file">Complete configuration file:</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content column-page-left" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Analysis of Canine TCR data from Zuleger et al 2024</h1>
</div>



<div class="quarto-title-meta column-page-left">

    
  
    
  </div>
  


</header>


<p>For this example, we assume that SatTCR is already downloaded, <code>snakemake</code> and <code>docker</code> are already installed in the system, and that the <code>docker</code> images are already pulled, i.e.&nbsp;we assume that the first steps of the <a href="./usage.html">Getting started</a> page are already done.</p>
<p>We also assume that we are located in the <code>SatTCR</code> directory after cloning the repository:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> clone git@github.com:Ong-Research/SatTCR</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> SatTCR</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="additional-files-to-download" class="level2">
<h2 class="anchored" data-anchor-id="additional-files-to-download">Additional files to download</h2>
<ol type="1">
<li>MIXCR’s license: This can be downloaded from their website: <a href="https://mixcr.com/mixcr/getting-started/milm/" class="uri">https://mixcr.com/mixcr/getting-started/milm/</a>. We rename the file as <code>license_mixcr</code>, and the name is coded in the <code>mixcr/license_file</code> field in the config file.</li>
<li>IMGT reference file: This can be downloaded from the <code>repseqio</code> repository release: <a href="https://github.com/repseqio/library-imgt/releases" class="uri">https://github.com/repseqio/library-imgt/releases</a>. In the screenshot below, we can see that the IMGT file could be <code>imgt.202214-2.sv8.json</code> after decompressing it.</li>
</ol>
<div class="cell">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="assets/repseqio_page.png" class="img-fluid figure-img" style="width:80.0%"></p>
<figcaption>Screenshot of the repseqio repository release page</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="making-the-samples.csv-file" class="level2">
<h2 class="anchored" data-anchor-id="making-the-samples.csv-file">Making the <code>samples.csv</code> file</h2>
<p>The name of this file is coded in the config file <code>samplefile</code> entry. For the moment, we assume that the sequencing files are located in the <code>data</code> folder. For example, in this repository, we have the following files:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> data</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="ex">data/run1/K9013v01-rxn1_S425_L003_R1_001.fastq.gz</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="ex">data/run1/K9013v01-rxn1_S425_L003_R2_001.fastq.gz</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="ex">data/run1/K9013v01-rxn2_S426_L003_R1_001.fastq.gz</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="ex">data/run1/K9013v01-rxn2_S426_L003_R2_001.fastq.gz</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>These are two paired RNA-seq files corresponding to two samples; so the <code>samples.csv</code> looks as:</p>
<pre class="csv"><code>sample_name,sample_file
run1_bio1_tech1,data/run1/K9013v01-rxn1_S425_L003
run1_bio1_tech2,data/run1/K9013v01-rxn2_S426_L003</code></pre>
<p>The first column is the name that is going to be used for the respective samples, and the second one is the name of the file until the “_R1” / “_R2” parts. The remaining part of the file names is defined in the <code>suffix</code> entry of the configuration file. In the case of these two files, that is <code>_001.fastq.gz</code>.</p>
</section>
<section id="running-the-pipeline" class="level2">
<h2 class="anchored" data-anchor-id="running-the-pipeline">Running the pipeline</h2>
<section id="quality-control" class="level3">
<h3 class="anchored" data-anchor-id="quality-control">Quality control</h3>
<p>The quality control module of the pipeline does the following:</p>
<ol type="1">
<li><p>Runs <code>fastqc</code> for every sequencing file, and gather all the results in a <code>multiqc</code> report.</p></li>
<li><p>Uses <code>dada2</code>’s to plot quality profiles.</p></li>
</ol>
<p>The underlying assumption for this step is the that <code>docker images</code> for the <code>fastqc</code>, <code>multiqc</code> and <code>rquarto</code> have been installed already. After this steps are completed, then run with 4 cores:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="at">-c4</span> qc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The quality profiles are saved in the <code>figs</code> directory, and one of them looks like:</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-example-qc-profile" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-example-qc-profile-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="cases/canine_tcr/figs/run1_bio1_tech1_qc_profile.png" class="img-fluid figure-img" style="width:80.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-example-qc-profile-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: Quallity profile for the tech1 files
</figcaption>
</figure>
</div>
</div>
</div>
<p>Each panel in the plot correspond to one of the sequence files for the sample:</p>
<ol type="1">
<li><p>The x-axis corresponds to the nucleotides in the sequences, labeled as <code>Cycles</code>, and the y-axis correspond to the quality score for each nucleotide in each sequence.</p></li>
<li><p>In the background of each panel, there is a heatmap indicating the % of nucleotides at the <span class="math inline">\(i\)</span>-th position among all sequences with a given quality score. The black bars at the top means that the majority of the nucleotides have a high score.</p></li>
<li><p>The green line is the avg. score at each position, where the orange line are the median (full) and 25% / 75% quantiles (dashed). These lines and the grey shades below 30 indicate that there are few sequences that need to be trimmed.</p></li>
</ol>
<p>Generally we could expected:</p>
<ul>
<li><p>The quality scores at the R2 end panel are a bit noisier than the ones in the R1 end panel.</p></li>
<li><p>The quality scores at the end of the sequences tend to be equal or lower than the ones a the start of the sequences.</p></li>
</ul>
</section>
<section id="trim-sequences" class="level3">
<h3 class="anchored" data-anchor-id="trim-sequences">Trim sequences</h3>
<p>This step assumes that the docker’s <code>trimmomatic</code> image has been already installed. Then, following the previous section, we know now that we need to trim the sequences. The tail at the end of the R2 panel indicates that there are a few sequences with QC score &lt; 20, so we remove the parts after that QC score is reached with the <code>TRAILING:20</code> option. There could be some remaining small fragments, and we remove the short fragments with length of 100 of less with the <code>MINLEN:100</code> option.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode yaml code-with-copy"><code class="sourceCode yaml"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">trim</span><span class="kw">:</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">trimmer</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"TRAILING:20"</span><span class="kw">,</span><span class="at"> </span><span class="st">"MINLEN:100"</span><span class="kw">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>More documentation on trimmomatic’s options are in <a href="http://www.usadellab.org/cms/?page=trimmomatic">their website</a>. We run this module with:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="at">-c4</span> trim</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="clonotype-assembly" class="level3">
<h3 class="anchored" data-anchor-id="clonotype-assembly">Clonotype assembly</h3>
<p>The main line to pay attention in the config file is the <code>mixcr/params</code> line. We used:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode yaml code-with-copy"><code class="sourceCode yaml"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="at">rna-seq --species dog -b imgt.202214-2 --rna</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>which means that our primers look line the <code>rna-seq</code> configuration from the MIXCR’s list of pre-set configurations. The species is <code>dog</code>, and the TCRB V and J gene sequences are obtained from the <code>imgt.202214-2</code> file (that we already downloaded and decompressed). MIXCR is very flexible, and <a href="https://mixcr.com/mixcr/reference/overview-mixins-list/">provides an extensive list of pre-set options</a>.</p>
<p>The clonotypes are assembled using the command:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="at">-c4</span> mixcr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>Note:</strong> There are other available species:</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Available species for TCR <span class="math inline">\(\beta\)</span> chain
</div>
</div>
<div class="callout-body-container callout-body">
<p>Human, Mouse, Ma’s night monkey, Rhesus Monkey , Rainbow trout, Dog, Ferret, Rabbit, Pig, Cat, Sheep, Camel, Crab-eating macaque, Naked mole-rat, Bovine, Mouse C57BL/6J, Gorilla</p>
</div>
</div>
</section>
<section id="sampling-of-the-sequencing-files-and-saturation-analysis" class="level3">
<h3 class="anchored" data-anchor-id="sampling-of-the-sequencing-files-and-saturation-analysis">Sampling of the sequencing files and saturation analysis</h3>
<p>There are two fields that are important for this section:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode yaml code-with-copy"><code class="sourceCode yaml"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">seed</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="dv">54232</span><span class="kw">,</span><span class="at"> </span><span class="dv">65432</span><span class="kw">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>and</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode yaml code-with-copy"><code class="sourceCode yaml"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saturation</span><span class="kw">:</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">samples</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"run1_bio1_tech1"</span><span class="kw">,</span><span class="at"> </span><span class="st">"run1_bio1_tech2"</span><span class="kw">]</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">bootstrap_replicates</span><span class="kw">:</span><span class="at"> </span><span class="dv">10</span><span class="co"> # 200  </span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co">  # only one of block_size or nblocks is supposed to be Null</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">block_size</span><span class="kw">:</span><span class="at"> </span><span class="ch">Null</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">nblocks</span><span class="kw">:</span><span class="at"> </span><span class="dv">10</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>This step samples pairs of sequences from the trimmed files corresponding to the <code>tech1</code> and <code>tech2</code> samples. The way it works is that for each random seed, it splits the paired sequence file into 10 blocks, in this case each block with 10% of the trimmed sequences, and then samples orderings among the 10 blocks. For example, an ordering could be <code>5, 4, 1, 2, 3, 6, 7, 8, 9, 10</code> and this will result with the 10% sequence file being block <code>5</code>, the 20% sequence file being blocks <code>5</code> and <code>4</code> together and so on.</p>
<p>In general, this steps generated <code>nblocks</code> <span class="math inline">\(\times\)</span> <code>bootstrap_replicates</code> <span class="math inline">\(\times\)</span> <code># of seeds</code> paired-sequence files for each one of the <code>samples</code> that were defined. This means that in our current scenario, we are generating <span class="math inline">\(2 \times 2 \times 10 \times 10 = 400\)</span> pairs of files.</p>
<p>This module is run using:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="at">-c4</span> sampling</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>For this step we suggest to use few seeds, and don’t divide the files into too many blocks. Then, to assembly the clonotypes of these files, we run the saturation module using:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="at">-c4</span> saturation</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="building-the-report" class="level3">
<h3 class="anchored" data-anchor-id="building-the-report">Building the report</h3>
<p>If the previous steps worked, then this is run using:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="at">-c4</span> report</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The <code>n_inter</code> field, is used to determine how many intersections to consider when plotting the # of shared clonotypes among samples.</p>
<p>The details of the report are coded in the <code>summary</code> section of the configuration file:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode yaml code-with-copy"><code class="sourceCode yaml"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span><span class="kw">:</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">min_count</span><span class="kw">:</span><span class="at"> </span><span class="dv">5</span><span class="co">    # min. # of times of repeats to consider a clonotype</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">vj_quantile</span><span class="kw">:</span><span class="at"> </span><span class="fl">.15</span><span class="co">    #  </span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">groups</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">"run1"</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"run1_bio1_tech1"</span><span class="kw">,</span><span class="at"> </span><span class="st">"run1_bio1_tech2"</span><span class="kw">,</span><span class="at"> </span><span class="st">"run1_bio2_tech1"</span><span class="kw">,</span><span class="at"> </span><span class="st">"run1_bio2_tech2"</span><span class="kw">],</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">"run2"</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"run2_bio1_tech1"</span><span class="kw">,</span><span class="at"> </span><span class="st">"run2_bio1_tech2"</span><span class="kw">,</span><span class="at"> </span><span class="st">"run2_bio2_tech1"</span><span class="kw">,</span><span class="at"> </span><span class="st">"run2_bio2_tech2"</span><span class="kw">],</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">"run3"</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"run3_bio1_tech1"</span><span class="kw">,</span><span class="at"> </span><span class="st">"run3_bio1_tech2"</span><span class="kw">,</span><span class="at"> </span><span class="st">"run3_bio2_tech1"</span><span class="kw">,</span><span class="at"> </span><span class="st">"run3_bio2_tech2"</span><span class="kw">]}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The <code>groups</code> part indicate which samples correspond to each group. In this case, the groups correspond to biological replicates, but it may well correspond to conditions or any other characteristics considered in the experiment.</p>
<p>The complete report is available in <strong>add link</strong>. For the saturation analysis, this will generate figures like <a href="#fig-saturation" class="quarto-xref">Figure&nbsp;2</a>. When the sequencing depth is too low, the growth is going to look approximately linear (like in the <code>run1</code> and <code>run2</code> panels), but if the # of sequences is adequate then the growth will show decreasing increments like in the <code>run3</code> panel.</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-saturation" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-saturation-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="cases/canine_tcr/saturation_files/figure-html/fig-saturation-nclones-1.png" class="img-fluid figure-img" style="width:80.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-saturation-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2: Saturation analysis for canine TCR control samples
</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="complete-configuration-file" class="level2">
<h2 class="anchored" data-anchor-id="complete-configuration-file">Complete configuration file:</h2>
<p>The complete <code>config.yaml</code> file looks like:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode yaml code-with-copy"><code class="sourceCode yaml"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="co"># general configuration files</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="fu">samplefile</span><span class="kw">:</span><span class="at"> </span><span class="st">"samples.csv"</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="fu">seed</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="dv">54232</span><span class="kw">,</span><span class="at"> </span><span class="dv">65432</span><span class="kw">]</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="fu">run_mixcr</span><span class="kw">:</span><span class="at"> </span><span class="ch">true</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="fu">run_saturation</span><span class="kw">:</span><span class="at"> </span><span class="ch">true</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="fu">run_report</span><span class="kw">:</span><span class="at"> </span><span class="ch">false</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="co"># The repertoire_assembly pipeline assumes that for a sample</span></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a><span class="co"># both end files have names of the form</span></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a><span class="co"># dict[sample] + "_R1" + {suffix}</span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a><span class="co"># dict[sample] + "_R2" + {suffix}</span></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a><span class="fu">suffix</span><span class="kw">:</span><span class="at"> </span><span class="st">"_001.fastq.gz"</span></span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a><span class="co"># docker images</span></span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a><span class="fu">docker</span><span class="kw">:</span><span class="at"> </span></span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">run_line</span><span class="kw">:</span><span class="at"> </span><span class="st">"docker run -v $(pwd):$(pwd) -w $(pwd) -u $(id -u):$(id -g)"</span></span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">fastqc</span><span class="kw">:</span><span class="at"> </span><span class="st">"staphb/fastqc"</span></span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">multiqc</span><span class="kw">:</span><span class="at"> </span><span class="st">"staphb/multiqc"</span></span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">trimmomatic</span><span class="kw">:</span><span class="at"> </span><span class="st">"staphb/trimmomatic"</span></span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">rquarto</span><span class="kw">:</span><span class="at"> </span><span class="st">"tcr/sat"</span></span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">mixcr</span><span class="kw">:</span><span class="at"> </span><span class="st">"ghcr.io/milaboratory/mixcr/mixcr:latest"</span></span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a><span class="co"># trimmomatic parameters</span></span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a><span class="fu">trim</span><span class="kw">:</span></span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">trimmer</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"TRAILING:20"</span><span class="kw">,</span><span class="at"> </span><span class="st">"MINLEN:100"</span><span class="kw">]</span></span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-29"><a href="#cb15-29" aria-hidden="true" tabindex="-1"></a><span class="fu">mixcr</span><span class="kw">:</span></span>
<span id="cb15-30"><a href="#cb15-30" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">params</span><span class="kw">:</span><span class="at"> </span><span class="st">"rna-seq --species dog -b imgt.202214-2 --rna"</span><span class="co"> # more presets can be seen https://docs.milaboratories.com/mixcr/reference/overview-built-in-presets/</span></span>
<span id="cb15-31"><a href="#cb15-31" aria-hidden="true" tabindex="-1"></a><span class="co">  # license_file: "/path/to/license/file"</span></span>
<span id="cb15-32"><a href="#cb15-32" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">license_file</span><span class="kw">:</span><span class="at"> </span><span class="st">"./license_mixcr"</span></span>
<span id="cb15-33"><a href="#cb15-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-34"><a href="#cb15-34" aria-hidden="true" tabindex="-1"></a><span class="fu">report</span><span class="kw">:</span></span>
<span id="cb15-35"><a href="#cb15-35" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">n_inter</span><span class="kw">:</span><span class="at"> </span><span class="dv">17</span></span>
<span id="cb15-36"><a href="#cb15-36" aria-hidden="true" tabindex="-1"></a><span class="at">  </span></span>
<span id="cb15-37"><a href="#cb15-37" aria-hidden="true" tabindex="-1"></a><span class="fu">saturation</span><span class="kw">:</span></span>
<span id="cb15-38"><a href="#cb15-38" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">samples</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"run1_bio1_tech1"</span><span class="kw">,</span><span class="at"> </span><span class="st">"run1_bio1_tech2"</span><span class="kw">,</span></span>
<span id="cb15-39"><a href="#cb15-39" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="st">"run2_bio1_tech1"</span><span class="kw">,</span><span class="at"> </span><span class="st">"run2_bio1_tech2"</span><span class="kw">,</span></span>
<span id="cb15-40"><a href="#cb15-40" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="st">"run3_bio1_tech1"</span><span class="kw">,</span><span class="at"> </span><span class="st">"run3_bio1_tech2"</span><span class="kw">]</span></span>
<span id="cb15-41"><a href="#cb15-41" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">bootstrap_replicates</span><span class="kw">:</span><span class="at"> </span><span class="dv">10</span><span class="co"> # 200  </span></span>
<span id="cb15-42"><a href="#cb15-42" aria-hidden="true" tabindex="-1"></a><span class="co">  # only one of block_size or nblocks is supposed to be Null</span></span>
<span id="cb15-43"><a href="#cb15-43" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">block_size</span><span class="kw">:</span><span class="at"> </span><span class="ch">Null</span></span>
<span id="cb15-44"><a href="#cb15-44" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">nblocks</span><span class="kw">:</span><span class="at"> </span><span class="dv">10</span></span>
<span id="cb15-45"><a href="#cb15-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-46"><a href="#cb15-46" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span><span class="kw">:</span></span>
<span id="cb15-47"><a href="#cb15-47" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">min_count</span><span class="kw">:</span><span class="at"> </span><span class="dv">5</span><span class="co">    # min. # of times of repeats to consider a clonotype</span></span>
<span id="cb15-48"><a href="#cb15-48" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">vj_quantile</span><span class="kw">:</span><span class="at"> </span><span class="fl">.15</span><span class="co">    #  </span></span>
<span id="cb15-49"><a href="#cb15-49" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">groups</span><span class="kw">:</span><span class="at"> </span><span class="kw">{</span></span>
<span id="cb15-50"><a href="#cb15-50" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">"run1"</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"run1_bio1_tech1"</span><span class="kw">,</span><span class="at"> </span><span class="st">"run1_bio1_tech2"</span><span class="kw">,</span><span class="at"> </span><span class="st">"run1_bio2_tech1"</span><span class="kw">,</span><span class="at"> </span><span class="st">"run1_bio2_tech2"</span><span class="kw">],</span></span>
<span id="cb15-51"><a href="#cb15-51" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">"run2"</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"run2_bio1_tech1"</span><span class="kw">,</span><span class="at"> </span><span class="st">"run2_bio1_tech2"</span><span class="kw">,</span><span class="at"> </span><span class="st">"run2_bio2_tech1"</span><span class="kw">,</span><span class="at"> </span><span class="st">"run2_bio2_tech2"</span><span class="kw">],</span></span>
<span id="cb15-52"><a href="#cb15-52" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">"run3"</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"run3_bio1_tech1"</span><span class="kw">,</span><span class="at"> </span><span class="st">"run3_bio1_tech2"</span><span class="kw">,</span><span class="at"> </span><span class="st">"run3_bio2_tech1"</span><span class="kw">,</span><span class="at"> </span><span class="st">"run3_bio2_tech2"</span><span class="kw">]}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="references" class="level2">
<h2 class="anchored" data-anchor-id="references">References</h2>
<ul>
<li><a href="https://doi.org/10.1016/j.vetimm.2023.110702">Zuleger CL, Welch Schwartz R, Ong IM, Newton MA, Vail DM, Albertini MR. <em>“Development of a next-generation sequencing protocol for the canine T cell receptor beta chain repertoire”</em>. Veterinary Immunology and Immunopathology (2024)</a></li>
</ul>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>Copyright (c) 2025 SMPH / Department of Biostatistics and Medical Informatics Ong Lab / UWCCC CISR</p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>